Applications/Spire/Source/BookView/BookViewController.cpp

```diff
@@ -36,7 +37,8 @@ void BookViewController::close() {
   if(m_window == nullptr) {
     return;
   }
-  m_window.reset();
+  m_window->deleteLater();
+  m_window = nullptr;
```

Comment:
It does not seem like these resets (the `nullptr` part) are always necessary, are they? It only makes a difference if there are explicit checks against `nullptr` somewhere afterwards.
Where necessary, you may consider wrapping these two statements into an inline function.

Reply:
Feel free to add a free function in Utility.hpp called ```deleteLater(QObject*& object)``` to do this.

Reply:
I think I understand the purpose of taking a reference to a pointer here (can reassign the original pointer), but how do I pass the object to this function? Passing in the window in BookViewController (type: BookViewWindow*) to this function gives me the error:

a reference of type "QObject*&" (not const-qualified) cannot be initialized with a value of type "Spire::BookViewWindow*"

The BookViewWindow is obviously derived from a QObject, so shouldn't the function parameter required it to be passed as a reference?

Reply:
Yes that's right, what I said is not legal or possible (bonus points if it makes sense to you why it's not possible). You will need to make it a template function.

    template<typename T>
    std::enable_if_t<std::is_base_of_v<QObject, T>> deleteLater(T*& object) {
      ...;
    }

Reply:
Is the issue potentially reassigning the reference?

Reply:
Yep, you got it! Very nice Sir.

Reply:
So how does the template version work, then? The template limits the types to derivatives of QObject, but they're still references to pointers.

Reply:
Okay sure, let's be more detailed about it then.  Imagine I had the following:

    void f(QObject*& object) {
      object = new QTimer(...);
    }

    void g() {
      auto q = new QScrollBar();
      f(q);
    }

Well it's clear why ```f(q)``` won't work, because ```f``` assigns to ```object``` a ```QTimer```, but we called ```f``` from within ```g``` with a ```QScrollBar```, so this is clearly not something that should be possible. Even though it's legal to assign a ```QTimer``` to a ```QObject```, it is not possible to assign a ```QTimer``` to a ```QScrollBar```.

However, if instead we had declared ```f``` as follows:

    void f(QObject*& const object);

Then this would be fine, because ```object``` can not be reassigned since it's ```const```, so there is no type safety violation.

Reply:
That makes sense.

Reply:
Why not just `void deleteLater(QObject** object)` ?

Reply:
Even with the reference version though, why do we need to explicitly disable the template for types that are not subclasses of `QObject`? Isn't it an example of type pessimization we normally avoid?

Reply:
I place a very high value on type safety. Once compiler support for C++20 and concepts becomes mainstream then there will be a major refactoring of generic code to be a lot more type safe than it currently is.